### PR TITLE
Extend Builtin Format function

### PR DESCRIPTION
**Pull request by staujd02**
Format was implemented into types where format(<type>, "") is the only accepted string format specifier

Format can now be implemented on more types.

## PR Checklist:
- [ x] All new features have been tested
- [ x] All new features have been documented
- [ x] I have read the **CONTRIBUTING.md** file
- [ x] I will abide by the code of conduct

### CODE REVIEW

File: tests/builtins/test_format.py [hunk 1]
```diff
@@ -1,9 +1,32 @@
 from .. utils import TranspileTestCase, BuiltinFunctionTestCase
 
-
 class FormatTests(TranspileTestCase):
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

File: tests/builtins/test_format.py [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
-    pass
+
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_format.py [hunk 3]
```diff
@@ -1,9 +1,32 @@
...
 class FormatTests(TranspileTestCase):
-    pass
+    
+    def test_format_throws_error_if_given_zero_args(self):
+        self.assertCodeExecution("""
+            try:
+                print(format())
+            except TypeError as error:
+                print(error)
+        """, run_in_function=False)
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py [hunk 4]
```diff
@@ -1,9 +1,32 @@
...
+        self.assertCodeExecution("""
+            try:
+                print(format("Too","Many","Args"))
+            except TypeError as error:
+                print(error)
+            try:
+                print(format("Way","Too","Many","Args"))
+            except TypeError as error:
+                print(error)
+        """, run_in_function=False)
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py [hunk 5]
```diff
@@ -1,9 +1,32 @@
...
+                print(format("Way","Too","Many","Args"))
+            except TypeError as error:
+                print(error)
+        """, run_in_function=False) 
 
+    def test_format_with_empty_string_for_second_arg_is_same_as_str(self):
+        self.assertCodeExecution("""
+            print(format("string", ""))
+            # print(format("string", "") == type("string").__format__("string",""))
+        """, run_in_function=False)
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py [hunk 6]
```diff
@@ -1,9 +1,32 @@
...
+                print(error)
+        """, run_in_function=False) 
 
+    def test_format_with_empty_string_for_second_arg_is_same_as_str(self):
+        self.assertCodeExecution("""
+            print(format("string", ""))
+            # print(format("string", "") == type("string").__format__("string",""))
+        """, run_in_function=False) 
 
 class BuiltinFormatFunctionTests(BuiltinFunctionTestCase, TranspileTestCase):
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

**Comment by martica** [comment]
Howdy @staujd02. Thanks for working on this!

The eslint check in CI is broken currently, so I'm stuck enforcing it manually at the moment. Can you replace the template strings with concatenation? Our .eslintrc currently doesn't allow them. :( You can refer to it in the top level of the repo.

File: tests/builtins/test_format.py (hunk 1 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

File: tests/builtins/test_format.py (hunk 2 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_format.py (hunk 3 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 4 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 5 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 6 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

File: tests/builtins/test_format.py [hunk 7]
```diff
@@ -2,8 +2,33 @@
 
 
 class FormatTests(TranspileTestCase):
-    pass
+
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_format.py [hunk 8]
```diff
@@ -2,8 +2,33 @@
...
+    def test_format_with_empty_string_for_second_arg_is_same_as_str(self):
+        self.assertCodeExecution("""
+            print(format("string", ""))
+            # print(format("string", "") == type("string").__format__("string",""))
+        """, run_in_function=False)
 
 
 class BuiltinFormatFunctionTests(BuiltinFunctionTestCase, TranspileTestCase):
-    function = "format"
+    function = "format"
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W391) blank line at end of file

**Comment by staujd02** [comment]
Sure thing @martica, Done! Took care of the auto linter too...

**Comment by martica** [comment]
I'll take a closer soon, but moving to calling the __format__ methods on types is a great step in the right direction here.

**Comment by martica** [comment]
Looks like there are a couple more errors that Brutus is grumbling about.

File: tests/builtins/test_format.py [hunk 9]
```diff
@@ -2,8 +2,33 @@
...
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W391) blank line at end of file

**Comment by staujd02** [comment]
With Vim's help I removed the last couple linter problems

**Comment by martica** [comment]
After a closer look, it feels like adding the stubs without any behaviour isn't gaining a lot yet. They all currently have `__format__` defined in their superclass, `object`, to just return the value for an empty format string, so defining stub methods doesn't provide any new functionality.

I think a more fruitful approach would be to implement `__format__` for one type at a time and make it useful. A lot of the functionality that should be in the `__format__` methods is in types/StrUtils.js and could be reused.

There is should be some decent examples in tests/datatypes/test_str.py for how to deal with formatting issues. You probably want to use the `@transforms` decorator on your tests or the current test infrastructure tries to reformat a number of things to aid with matching, but the output of format() should match without that. @awildtechno just wrote some docs on how to use that: https://github.com/beeware/batavia/blob/8c55a01a06687ee54fe5f5e37292571a3aa3c399/tests/utils/output_cleaners.py#L25

**Comment by martica** [comment]
If you want to chat a bit about how to best approach this, come join us on gitter (https://gitter.im/beeware/general).

**Comment by staujd02** [comment]
Thanks for the review @martica , and for informing me about the transforms decorator. During some exploratory testing I noticed that reformatting behavior (but was unaware on how to stop it).

If I fully implement the __format__ option for the Bool type, would it be best to start a new pull request or would preserving the context of this pull request be preferable?

**Comment by martica** [comment]
You can start a new PR, or continue this one. Either way is fine.

Format has a lot of functionality. Don't feel like you have to completely implement Bool. Any additional functionality is valuable.

File: tests/builtins/test_format.py (hunk 1 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

File: tests/builtins/test_format.py (hunk 2 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_format.py (hunk 3 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 4 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 5 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 36: (W291) trailing whitespace

File: tests/builtins/test_format.py (hunk 6 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E302) expected 2 blank lines, found 1

File: tests/builtins/test_format.py (hunk 9 above)
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W391) blank line at end of file

**Comment by awildtechno** [comment]
Looking at this briefly, I think most of the functionality already exists for you in Str.format. Could you do something like this?

s = "{" + formatSpecifier + "}"
return callables.call_method(s, 'format', [this], {})  # Batavia equivalent of s.format(this)

Does that make sense?

**Comment by staujd02** [comment]
Yep, that makes sense.

**Comment by martica** [comment]
We do need to keep in mind that eventually `str.format` should be calling `__format__` on the types that are being formatted. For now, having the various `__format__` methods call `StrUtils._new_subsitute` would get a lot of reuse quickly.

For a long-term goal, it feels like we need to find a seem in `StrUtils._new_subsitute` between its parsing of the full format strings and its formatting of individual values and `__format__` in the types could call into the latter part.

**Comment by staujd02** [comment]
This is proving to be problematic. StrUtils is converting my Boolean into a string every time, so a call to format(True, 'f') --> bool._._format_._(True, 'f') ---> ("{:" + "f" + "}").format( True )
yields Unknown format code 'f' for object of type 'str'
rather than 1.00000

**Comment by staujd02** [comment]
I'm trying to redirect the formatting to `string.format()` and in effect StringUtils.__new_substitute(). But the arguments are being converted into a string types and therefore type conversions valid for the Boolean type are being rejected since they don't apply to strings...

**Comment by martica** [comment]
That sounds frustrating. I can try to help this evening (PDT) if you are still stuck.

**Comment by awildtechno** [comment]
@martica so if I'm understanding right, the end result should actually be Str.format calls something like <type>.__format__(arg, format_style) for each substitution? That's going to be a very, very thick commit as the current format is about 1500 lines of switch/case choose your own adventure.

**Comment by martica** [comment]
That's the way cpython implements it, or at least makes it appear that way.
If you implement `__format__` on a custom type string.format will call it.
We don't have to follow the exact implementation, but does make some things
easier.

That nest of switches is something that would be much more clearly
represented with polymorphism.

**Comment by martica** [comment]
I took a look @staujd02. It looks like the problem comes in here: https://github.com/beeware/batavia/blob/master/batavia/types/StrUtils.js#L1173

Bool is actually an outlier in that if you ask for string formatting, it shows the english word, but any numeric formatting treats it as 0 or 1. Seems like that chain of ifs needs a case for Bool, and because it depends on the format type (empty string or 's' vs. all the other types) you'll need to move the format parsing before the arg gets set. Look at https://github.com/beeware/batavia/blob/master/batavia/types/StrUtils.js#L796

Fixing the current string.format() is a major refactoring. Don't worry about that. Having the format builtin call string.format directly should get us a lot without making the future refactor any harder.

**Comment by staujd02** [comment]
@martica that sounds like a plan. I'll bump the PR if something else crops up.

File: tests/datatypes/test_bool.py [hunk 10]
```diff
@@ -1,8 +1,94 @@
+import unittest
 from ..utils import TranspileTestCase, UnaryOperationTestCase, BinaryOperationTestCase, InplaceOperationTestCase, \
-    MagicMethodFunctionTestCase
+    MagicMethodFunctionTestCase, transforms, adjust
+from itertools import product
 
 
 class BoolTests(TranspileTestCase):
+
```
Comment (brutusthebee) [inline tests/datatypes/test_bool.py]: At column 1: (W293) blank line contains whitespace

File: tests/datatypes/test_bool.py [hunk 11]
```diff
@@ -1,8 +1,94 @@
...
+    @unittest.expectedFailure
+    @transforms(
+        decimal=False,
+        float_exp=False,
+        high_precision_float=False,
+        js_bool=False,
+        complex_num=False)
+    def test_character_conversion_type(self, js_cleaner, py_cleaner):
+        self.assertCodeExecution( """
```
Comment (brutusthebee) [inline tests/datatypes/test_bool.py]: At column 4: (W291) trailing whitespace

Comment (brutusthebee) [inline tests/datatypes/test_bool.py]: At column 34: (E201) whitespace after '('

File: tests/datatypes/test_bool.py [hunk 12]
```diff
@@ -1,8 +1,94 @@
...
+        decimal=False,
+        float_exp=False,
+        high_precision_float=False,
+        js_bool=False,
+        complex_num=False)
+    def test_character_conversion_type(self, js_cleaner, py_cleaner):
+        self.assertCodeExecution( """ 
+            print(format(True, 'c'))
+            print(format(False, 'c'))
+            """ , js_cleaner=js_cleaner, py_cleaner=py_cleaner)
```
Comment (brutusthebee) [inline tests/datatypes/test_bool.py]: At column 16: (E203) whitespace before ','

File: tests/datatypes/test_bool.py [hunk 13]
```diff
@@ -1,8 +1,94 @@
...
+        js_bool=False,
+        decimal=False,
+        float_exp=False,
+        memory_ref=False
+    )
+    def test_simple_format(self, js_cleaner, py_cleaner):
+        self.assertCodeExecution("""
+            print(format(True, ''))
+            """, js_cleaner=js_cleaner, py_cleaner=py_cleaner)
+
```
Comment (brutusthebee) [inline tests/datatypes/test_bool.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_format.py [hunk 14]
```diff
@@ -1,8 +1,47 @@
-from .. utils import TranspileTestCase, BuiltinFunctionTestCase
+from .. utils import TranspileTestCase, BuiltinFunctionTestCase, transforms
+
 
 
 class FormatTests(TranspileTestCase):
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (E303) too many blank lines (3)

File: tests/builtins/test_format.py [hunk 15]
```diff
@@ -1,8 +1,47 @@
...
+            try:
+                print(format("Too","Many","Args"))
+            except TypeError as error:
+                print(error)
+            try:
+                print(format("Way","Too","Many","Args"))
+            except TypeError as error:
+                print(error)
+        """, run_in_function=False)
+
```
Comment (brutusthebee) [inline tests/builtins/test_format.py]: At column 1: (W293) blank line contains whitespace

**Comment by staujd02** [comment]
Thanks for the help @martica and @awildtechno ! Hope this is close to what you are looking for...

**Comment by awildtechno** [comment]
I'm not really qualified to review but I wanted to say Thanks for sticking it out with this. I just got started here a few weeks ago with some modifications to Str.format and I know how nitpicky writing and passing tests for the formatter can be.

**Review by phildini — APPROVED** [review]

**Pull request merged**